[query] New Shuffler

johnc1231 opened the pull request:
PRing for test suite, but it's mostly working. 

Todo:

- [x] Optimization for already sorted tables
- [x] ~~Handle sort by descending~~ (deferred to subsequent PR)
- [x] Handle tables with no partitions

johnc1231 commented:
I assigned both Tim and Patrick for this since we've been talking about it. It still has prints and stuff in it that I'm using to debug / and experiment, but I think it's now in a place where a review would be welcome.

patrick-schultz reviewed (changes requested):
Mostly looks good, though I found it quite hard to read. A few more comments in the big sort method might help.

johnc1231 commented on hail/src/main/scala/is/hail/expr/ir/lowering/LowerDistributedSort.scala:
```diff
@@ -78,4 +76,407 @@ object LowerDistributedSort {
+    rowTypeRequiredness: RStruct
+  ): TableStage = {
+
+    val oversamplingNum = 1
```
Ah, sorry. Yeah I have to increase, right now we are basically not using this.

johnc1231 commented:
@patrick-schultz I addressed your comments, would be happy to take more steps to make code clearer if you see some. Otherwise I'd like if you could check that you're now happy with how `oversamplingNum` is being used.

patrick-schultz commented on hail/src/main/scala/is/hail/expr/ir/lowering/LowerDistributedSort.scala:
```diff
@@ -176,17 +176,22 @@ object LowerDistributedSort {
-              "intervalTuple" -> MakeTuple.ordered(Seq(GetField(aggResults, "min"), GetField(aggResults, "max")))
-            ))
+            bindIR(sortedOversampling) { sortedOversampling =>
+              val sortedSampling = ToArray(mapIR(StreamRange(0, ArrayLen(sortedOversampling), I32(oversamplingNum))) { idx =>
```
I believe the range you want is `range(oversamplingNum - 1, length, oversamplingNum)`. For example, say `branchingFactor = 4` and `oversamplingNum = 3`. Then we collect `4*3-1 = 11` "oversamples", and we want to extract `4-1` evenly spaced samples:
```
[*, *, 2, *, *, 5, *, *, 8, *, *]
```

patrick-schultz reviewed (changes requested):
With one fix to the oversampling logic (and getting tests passing), I'm good with merging this.

johnc1231 commented:
Ok Patrick, back to you. I realized `PruneDeadFields` was disabled in my PR, I renabled it, and had to adjust the prune rule for `AggFold`.

patrick-schultz reviewed (approved):
Sorry I've been so slow reviewing, this is great work!
